uptime/cached_state.go

```diff
@@ -0,0 +1,103 @@
+package uptime
+
+import (
+	"sync"
+	"context"
+
+	"github.com/filecoin-project/lotus/api/v0api"
+	"github.com/filecoin-project/go-address"
+)
+
+type InnerState = HAMTState
+
+type CacheState struct {
+	self ActorID
+
+	inner InnerState
+	
+	// Local cache of voted checkers. TODO: use LRU cache
+	processedCheckers map[ActorID]bool
+
+	rwLock sync.RWMutex
+}
+
+func Load(ctx context.Context, api v0api.FullNode, actorAddr address.Address, self ActorID) (CacheState, error)  {
+	s, err := LoadHAMTState(ctx, api, actorAddr)
+	if err != nil {
+		return CacheState{}, err
+	}
+	return CacheState {
+		self: self,
+		inner: s,
+		processedCheckers: make(map[ActorID]bool),
+	}, nil
+}
+
+func (c *CacheState) HasRegistered(actor ActorID) (bool, error) {
+	return c.inner.HasRegistered(actor)
+}
+
+func (c *CacheState) ListReportedCheckerNotVoted() (map[ActorID]*[]MultiAddr, error) {
+	ids := make(map[ActorID]*[]MultiAddr)
+	
+	l, err := c.inner.GetOfflineCheckers()
+	if err != nil {
+		return ids, err
+	}
+
+	log.Errorw("list of offline checkers", "checkers", l)
+
+	for _, actorID := range(l) {
+		hasVoted, err := c.HasVotedReportedPeer(actorID)
+		if err != nil {
+			return ids, err
```

Comment:
Probably better to `return nil, err` to force a nil pointer error in case someone decides to disregard the `err`.